runtime/gc_glue_java/ObjectModel.hpp

```diff
@@ -588,12 +578,58 @@ class GC_ObjectModel : public GC_ObjectModelBase
-			} else {
-				hashOffset = _mixedObjectModel->getHashcodeOffset(clazz);
-			}
+			hashOffset = _mixedObjectModel->getHashcodeOffset(clazz);
+			uint32_t *hashCodePointer = (uint32_t*)((uint8_t*) destinationObjectPtr + hashOffset);
+			*hashCodePointer = convertValueToHash(_javaVM, (uintptr_t)forwardedHeader->getObject());
+			setObjectJustHasBeenMoved(destinationObjectPtr);
+		}
+	}
+
+	/**
+	 * Update the new version of this indexable object after it has been copied. This undoes any damaged
+	 * caused by installing the forwarding pointer into the original prior to the copy, and sets
+	 * the object age.
+	 *
+	 * This will install the correct (i.e. unforwarded) class pointer, update the hashed/moved
+	 * flags and install the hash code if the indexable object has been hashed but not previously moved.
+	 *
+	 * @param[in] forwardedHeader pointer to the MM_ForwardedHeader instance encapsulating the indexable object
+	 * @param[in] destinationObjectPtr pointer to the copied indexable object to be fixed up
+	 * @param[in] objectAge the age to set in the copied indexable object
+	 */
+	MMINLINE void
+	fixupForwardedObjectIndexable(MM_ForwardedHeader *forwardedHeader, omrobjectptr_t destinationObjectPtr, uintptr_t objectAge)
+	{
+		GC_ObjectModelBase::fixupForwardedObject(forwardedHeader, destinationObjectPtr, objectAge);
 
+		/* Updates internal field of indexable objects. Every indexable object have an extra field
+			* that can be used to store any extra information about the indexable object. One use case is
+			* OpenJ9 where we use this field to point to array data. In this case it will always point to
+			* the address right after the header, in case of contiguous data it will point to the data
+			* itself, and in case of discontiguous arraylet it will point to the first arrayiod. How to
+			* updated dataAddr is up to the target language that must override fixupDataAddr */
+		_indexableObjectModel->fixupDataAddr(forwardedHeader, destinationObjectPtr);
+
+		/*	To have ability to backout last scavenge we need to recognize objects just moved (moved first time) in current scavenge
+		 *
+		 *	Bits	State
+		 *	--------------------------------
+		 * 	m h		moved / hashed bits
+		 * 	0 0		not moved / not hashed
+		 * 	0 1		not moved / hashed
+		 * 	1 0		just moved / hashed
+		 * 	1 1		moved / hashed
+		 *	--------------------------------
+		 */
+		if (hasBeenMoved(getPreservedFlags(forwardedHeader))) {
+			if (hasRecentlyBeenMoved(getPreservedFlags(forwardedHeader))) {
+				/* Moved bit set / hashed bit not set means "moved previous scavenge" so set moved/hashed */
+				setObjectHasBeenMoved(destinationObjectPtr);
+			}
+		} else if (hasBeenHashed(getPreservedFlags(forwardedHeader))) {
+			/* The object has been hashed and has not been moved so we must store the previous address into the hashcode slot at hashcode offset. */
+			uintptr_t hashOffset;
+			hashOffset = _indexableObjectModel->getPreservedHashcodeOffset(forwardedHeader);
```

Comment:
Let's continue this route of having specialized mixed/indexable APIs and see how it looks and performs, but let's also try to share hashing code between them (something like fixupHashingFlagsAndSlot). That will require resurrection of isIndexable check in that new method, but I'd be ok with it since, it's infrequently executed (only on first movement of a hashed object).